src/Components/Components/src/Routing/TemplateParser.cs

```diff
@@ -13,21 +13,20 @@ namespace Microsoft.AspNetCore.Components.Routing
     // simple parameters from it.
     // Some differences with ASP.NET Core routes are:
     // * We don't support catch all parameter segments.
-    // * We don't support optional parameter segments.
     // * We don't support complex segments.
     // The things that we support are:
     // * Literal path segments. (Like /Path/To/Some/Page)
     // * Parameter path segments (Like /Customer/{Id}/Orders/{OrderId})
     internal class TemplateParser
     {
         public static readonly char[] InvalidParameterNameCharacters =
-            new char[] { '*', '?', '{', '}', '=', '.' };
+            new char[] { '*', '{', '}', '=', '.' };
 
         internal static RouteTemplate ParseTemplate(string template)
```

Comment:
This PR might be missing support for optional parameters that do not have constraints, for .e.g a template that looks like: `/Customers/{id?}`. If that isn't already supported, you might have to make `IsOptional` a feature of `TemplateSegment` (similar to `TemplateSegment.IsParameter`) and use that to make decisions.

Reply:
Should something in the parser enforce that optional parameters cannot be followed by non-optional ones? I was kinda surprised routing in the server doesn't really complain about this, but those parameters basically end up being treated as required

Reply:
> This PR might be missing support for optional parameters that do not have constraints, for .e.g a template that looks like: /Customers/{id?}.

Yes, Steve brought this up and I started noodling on it.

>  If that isn't already supported, you might have to make IsOptional a feature of TemplateSegment (similar to TemplateSegment.IsParameter) and use that to make decisions.

That's what I started implementing yesterday so good to know I am on the right track with this. I'm currently refactoring the PR to use `TemplateSegment.IsOptional` more heavily and will push the update commits soon.

> Should something in the parser enforce that optional parameters cannot be followed by non-optional ones?

Yeah, I think the fact that routing on the server allows you to create templates where non-optional parameters follow optional ones is a bit of a code smell. I can add logic in the parser to throw an exception when this happens.

Reply:
I've just pushed the commit that adds the `IsOptional` parameter to `TemplateSegment`.